>Integration tests (asan) [4/6] — fail: 1, passed: 374, flaky: 0
>Integration tests (release) [4/4] — fail: 1, passed: 525, flaky: 0
>Integration tests (tsan) [4/6] — fail: 1, passed: 374, flaky: 0


- `test_extreme_deduplication` - #52085 

>Integration tests (asan, analyzer) [1/6] — fail: 34, passed: 261, flaky: 0

- Differences in exception names - I.e. `Unknown function MySum` vs `Function with name 'MySum' does not exists`
- Somewhere the server crashes (since optimize_skip_unused_shards leads to SIGSEGV now, I was going to look at this after https://github.com/ClickHouse/ClickHouse/pull/51037)
- `TCPHandler: Code: 10. DB::Exception: Not found column tcpPort() in block.`
- `2023.07.13 13:39:08.509341 [ 8 ] {} <Error> Application: DB::Exception: Setting allow_experimental_analyzer is neither a builtin setting nor started with the prefix 'custom_' registered for user-defined settings`

>Integration tests (release) [2/4] — fail: 8, passed: 565, flaky: 0

- `test_lost_part`

```
E           AssertionError: Still have something in replication queue:
E             Row 1:
E             ──────
E             database:               default
E             table:                  mt3
E             replica_name:           node2
E             position:               0
E             node_name:              queue-0000000000
E             type:                   GET_PART
E             create_time:            2023-07-13 14:07:07
E             required_quorum:        0
E             source_replica:         node1
E             new_part_name:          8b8f0fede53df97513a9fb4cb19dc1e4_0_0_0
E             parts_to_merge:         []
E             is_detach:              0
E             is_currently_executing: 0
E             num_tries:              13108
E             last_exception:         
E             last_exception_time:    1970-01-01 00:00:00
E             last_attempt_time:      2023-07-13 14:13:07
E             num_postponed:          13593
E             postpone_reason:        Not executing log entry queue-0000000000 for part 8b8f0fede53df97513a9fb4cb19dc1e4_0_0_0 because it is not disjoint with part 8b8f0fede53df97513a9fb4cb19dc1e4_0_0_0_1 that is currently executing.
E             last_postpone_time:     2023-07-13 14:13:07
E             merge_type:             
E             
E             Row 2:
E             ──────
E             database:               default
E             table:                  mt3
E             replica_name:           node2
E             position:               1
E             node_name:              queue-0000000001
E             type:                   MUTATE_PART
E             create_time:            2023-07-13 14:07:07
E             required_quorum:        0
E             source_replica:         node1
E             new_part_name:          8b8f0fede53df97513a9fb4cb19dc1e4_0_0_0_1
E             parts_to_merge:         ['8b8f0fede53df97513a9fb4cb19dc1e4_0_0_0']
E             is_detach:              0
E             is_currently_executing: 1
E             num_tries:              13108
E             last_exception:         
E             last_exception_time:    1970-01-01 00:00:00
E             last_attempt_time:      2023-07-13 14:13:07
E             num_postponed:          13107
E             postpone_reason:        Not executing log entry queue-0000000001 for part 8b8f0fede53df97513a9fb4cb19dc1e4_0_0_0_1 because it is not disjoint with part 8b8f0fede53df97513a9fb4cb19dc1e4_0_0_0 that is currently executing.
E             last_postpone_time:     2023-07-13 14:13:07
E             merge_type:             
E             
E           assert False
```

>Integration tests (tsan) [5/6] — fail: 1, passed: 521, flaky: 2

- `test_storage_s3/test.py::test_url_reconnect_in_the_middle - Exception: ...` - https://github.com/ClickHouse/ClickHouse/issues/51345

--- a/.gitignore
+++ b/.gitignore
@@ -161,6 +161,7 @@ tests/queries/0_stateless/test_*
 tests/queries/0_stateless/*.binary
 tests/queries/0_stateless/*.generated-expect
 tests/queries/0_stateless/*.expect.history
+tests/integration/**/_gen
 
 # rust
 /rust/**/target

--- a/docker/test/integration/runner/Dockerfile
+++ b/docker/test/integration/runner/Dockerfile
@@ -135,4 +135,5 @@ ENV MSAN_OPTIONS='abort_on_error=1 poison_in_dtor=1'
 
 EXPOSE 2375
 ENTRYPOINT ["dockerd-entrypoint.sh"]
-CMD ["sh", "-c", "pytest $PYTEST_OPTS"]
+# To pass additional arguments (i.e. list of tests) use PYTEST_ADDOPTS
+CMD ["sh", "-c", "pytest"]

--- a/tests/integration/ci-runner.py
+++ b/tests/integration/ci-runner.py
@@ -10,6 +10,7 @@ import random
 import shutil
 import subprocess
 import time
+import shlex
 import zlib  # for crc32
 
 
@@ -646,7 +647,7 @@ class ClickhouseIntegrationTestsRunner:
             info_basename = test_group_str + "_" + str(i) + ".nfo"
             info_path = os.path.join(repo_path, "tests/integration", info_basename)
 
-            test_cmd = " ".join([test for test in sorted(test_names)])
+            test_cmd = " ".join([shlex.quote(test) for test in sorted(test_names)])
             parallel_cmd = (
                 " --parallel {} ".format(num_workers) if num_workers > 0 else ""
             )

--- a/tests/integration/runner
+++ b/tests/integration/runner
@@ -11,6 +11,7 @@ import subprocess
 import sys
 import string
 import random
+import shlex
 
 
 def random_str(length=6):
@@ -407,8 +408,14 @@ if __name__ == "__main__":
     if args.analyzer:
         use_analyzer = "-e CLICKHOUSE_USE_NEW_ANALYZER=1"
 
-    pytest_opts = " ".join(args.pytest_args).replace("'", "\\'")
-    tests_list = " ".join(args.tests_list)
+    # NOTE: since pytest options is in the argument value already we need to additionally escape '"'
+    pytest_opts = " ".join(
+        map(lambda x: shlex.quote(x).replace('"', '\\"'), args.pytest_args)
+    )
+    tests_list = " ".join(
+        map(lambda x: shlex.quote(x).replace('"', '\\"'), args.tests_list)
+    )
+
     cmd_base = (
         f"docker run {net} {tty} --rm --name {CONTAINER_NAME} "
         "--privileged --dns-search='.' "  # since recent dns search leaks from host
@@ -420,7 +427,7 @@ if __name__ == "__main__":
         f"--volume={args.src_dir}/Server/grpc_protos:/ClickHouse/src/Server/grpc_protos "
         f"--volume=/run:/run/host:ro {dockerd_internal_volume} {env_tags} {env_cleanup} "
         f"-e DOCKER_CLIENT_TIMEOUT=300 -e COMPOSE_HTTP_TIMEOUT=600 {use_analyzer} -e PYTHONUNBUFFERED=1 "
-        f"-e PYTEST_OPTS='{parallel_args} {pytest_opts} {tests_list} {rand_args} -vvv'"
+        f'-e PYTEST_ADDOPTS="{parallel_args} {pytest_opts} {tests_list} {rand_args} -vvv"'
         f" {DIND_INTEGRATION_TESTS_IMAGE_NAME}:{args.docker_image_version}"
     )
 

--- a/tests/integration/runner
+++ b/tests/integration/runner
@@ -136,9 +136,7 @@ def check_args_and_update_paths(args):
 
 def docker_kill_handler_handler(signum, frame):
     subprocess.check_call(
-        'docker ps --all --quiet --filter name={name} --format="{{{{.ID}}}}"'.format(
-            name=CONTAINER_NAME
-        ),
+        "docker ps --all --quiet --filter name={name}".format(name=CONTAINER_NAME),
         shell=True,
     )
     raise KeyboardInterrupt("Killed by Ctrl+C")
@@ -438,7 +436,7 @@ if __name__ == "__main__":
     )
 
     containers = subprocess.check_output(
-        f"docker ps --all --quiet --filter name={CONTAINER_NAME} --format={{{{.ID}}}}",
+        f"docker ps --all --quiet --filter name={CONTAINER_NAME}",
         shell=True,
         universal_newlines=True,
     ).splitlines()

